snuba/migrations/snuba_migrations/metrics/0020_polymorphic_buckets_table.py

```diff
@@ -0,0 +1,74 @@
+from typing import Sequence
+
+from snuba.clusters.storage_sets import StorageSetKey
+from snuba.migrations import migration, operations, table_engines
+from snuba.migrations.columns import MigrationModifiers as Modifiers
+from snuba.utils.schemas import Array, Column, DateTime, Float, Nested, String, UInt
+
+
+class Migration(migration.ClickhouseNodeMigration):
+    """
+    Create a polymorphic bucket table for metrics_sets,
+    metrics_distributions, and metrics_counters
+    """
+
+    blocking = False
+    local_table_name = "metrics_buckets_consolidated"
+    dist_table_name = "metrics_buckets_consolidated_dist"
+
+    column_list: Sequence[Column[Modifiers]] = [
+        Column("org_id", UInt(64)),
+        Column("project_id", UInt(64)),
+        Column("metric_id", UInt(64)),
+        Column("timestamp", DateTime()),
+        Column("tags", Nested([Column("key", UInt(64)), Column("value", UInt(64))]),),
+        Column("metric_type", String(Modifiers(low_cardinality=True))),
+        Column("set_values", Array(UInt(64))),
```

Comment:
I wonder whether empty array or null will be more efficient to store.

Reply:
You cannot make arrays nullable (https://clickhouse.com/docs/en/sql-reference/data-types/nullable/#data_type-nullable) but an empty array seems to take about as much space as a scalar Float64 (see https://gist.github.com/onewland/20799a8feee3317a1a6f6061f28297a9 for some testing I did around NULL)